File: cpp/1442_形成两个异或相等数组的三元组数目_medium.cpp
class Solution {
public:
    int countTriplets(vector<int>& arr) {
        int n = arr.size();
        int ans = 0;
        vector<int> s(n+1);
        for (int i=0; i<n; i++)
            s[i+1] = s[i] ^ arr[i];
        unordered_map<int, int> cnt, total;
        for (int k=0; k<n; k++) {
            if (cnt[s[k+1]])
                ans += cnt[s[k+1]] * k - total[s[k+1]];
            ++cnt[s[k]];
            total[s[k]] += k;
        }
        return ans;
    }
};

File: python/1442_形成两个异或相等数组的三元组数目_medium.py
class Solution:
    def countTriplets(self, arr: List[int]) -> int:
        ans, n, s = 0, len(arr), [0]
        for i in range(n):
            s.append(s[-1] ^ arr[i])
        cnt, total = Counter(), Counter()
        for k in range(n):
            if s[k+1] in cnt:
                ans += cnt[s[k+1]] * k - total[s[k+1]]
            cnt[s[k]] += 1
            total[s[k]] += k
        return ans